File: python/tradingflow/portfolio/mean/proportional.py
import numpy as np

from .._base import MeanPortfolio, mean_portfolio


def weights_of(mu: np.ndarray) -> np.ndarray:
    """Weight proportional to predicted return, shorting nothing."""
    weights = np.maximum(mu, 0.0)
    total = weights.sum()
    return weights / total if total > 0 else weights


def build(**kwargs) -> MeanPortfolio:
    """Constructs a proportional-weight portfolio.

    Stocks with a non-positive predicted return get nothing; the rest split the
    book in proportion to what they are expected to return. Conviction-weighted
    and trivially cheap, but with no risk model it will happily concentrate the
    whole portfolio into one volatile name that happens to score highest.
    """
    return mean_portfolio(weights_of, **kwargs)

File: crates/tradingflow/src/operators/portfolio/mean/proportional.rs
use super::MeanPortfolio;
use crate::operators::portfolio::Config;
use crate::python::py_operator_module;

/// Weights each stock in proportion to its predicted return.
///
/// Stocks predicted to fall get nothing; the rest split the book by how much
/// they are expected to return. The most direct translation of a prediction
/// into a position, and the one most exposed to a bad one — with no risk model
/// it will concentrate the whole book into a single volatile name that happens
/// to score highest.
///
/// See [module-level docs](super::super) for inputs and outputs.
pub fn proportional(config: Config) -> impl MeanPortfolio {
    py_operator_module(
        "tradingflow.portfolio.mean.proportional",
        config.params(|_| Ok(())),
    )
}
